Automattic/wp-calypso
# Fix SIWA when users have 2FA enabled

### Changes proposed in this Pull Request

This PR fixes signing in with Apple when 2FA is enabled.

What was happening is that at the last stage of the login flow, in [LoginBlock](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/login/index.jsx#L425), there are two boolean flags called `twoFactorEnabled` and `socialConnect`.

These flags, when `true`, are used to asynchronously fetch and render respective login blocks. The issue is that LoginBlock renders said blocks in an either or fashion. If `twoFactorEnabled` is enabled, the component will return the 2FA login block and bail out. Making `socialConnect` flag useless.

So if the user has 2FA **AND** Social login enabled, the social login block will not be fetched nor rendered.

This PR flips the order and makes `socialConnect` flag take precedence. Because this flag is only set to `true` after the 2FA flow is already done, so there is no need to render the 2FA block by the time `socialConnect` is set `true`.

I know this because `socialConnect` [is only set to `true` when the user is redirected to `social-connect`](https://github.com/Automattic/wp-calypso/blob/master/client/login/controller.js#L41) path. Whereas `twoFactorEnabled` remains `true` as long as the user is identified to have 2FA enabled. 

### Testing

Automatticians can't use social login. And my only Apple ID is an a8c account. So I can't test it. But the hiccup was before the a8c check, so I know this PR at least fixes the current faulty point. But I couldn't test beyond that.

Apple will not accept `calypso.live` domain as a valid domain for "sign in with Apple". To work around that please follow the following steps:

1. Checkout this branch locally and start Calypso. This PR has special config and will accept HTTPS requests on port 443.

2. Add `127.0.0.1 wordpress.com` to your hosts file.

2. Then visit https://wordpress.com/log-in. You should see a certificate warning. In this Chrome warning screen, type "thisisunsafe". This is a cheat code for Chrome that allows you to ignore certificate issues.

3. Once the page loads, please follow [these instructions](https://github.com/Automattic/wp-calypso/issues/42359#:~:text=Steps%20to%20reproduce) and see if the issue is still reproducible. 


Fixes #42359
